### ISSUE TITLE
Problem connecting with winpty

### ISSUE THREAD

**Issue by lperamo** [tiangolo/babun-docker#4]
When I launch `docker run-it -v $(pwd):/c/Users/dev2/Projets\ web default bash` it shows me that :

> cannot enable tty mode on non tty input
> babun-docker: Using winpty
> Post http://127.0.0.1:2375/v1.20/containers/create: dial tcp 127.0.0.1:2375: ConnectEx tcp: Aucune connexion n'a pu être établie car l'ordinateur cible l'a expressément refusée..
> - Are you trying to connect to a TLS-enabled daemon without TLS?
> - Is your docker daemon up and running?

`docker ps` show me that :

> Get http://127.0.0.1:2375/v1.20/containers/json: dial tcp 127.0.0.1:2375: ConnectEx tcp: Aucune connexion n’a pu être établie car l’ordinateur cible l’a expressément refusée..
> babun-docker: Trying to start docker-machine default
> Started machines may have new IP addresses. You may need to re-run the `docker-machine env` command.
> babun-docker: Setting up docker-machine environment
> open C:\Users\dev2.docker\machine\machines\default\server.pem: Le fichier spécifié est introuvable.
> babun-docker: Running command again

I'm under Windows 8, Babun launched via ConEmu.

**Comment by xinity** [comment]
hi,

i would advise you to first run your container daemonized (-d flag) and the do : 
docker exec -it [your container] /bin/bash

your issue seems to be related to DOCKER_HOST env variable using accessible by winpty

hoping this helps,

**Comment by lperamo** [comment]
`docker run -d -v $(pwd):/c/Users/dev2/Projets\ web default bash`
and then
`docker exec -it [your container] /bin/bash`
?

**Comment by xinity** [comment]
docker run -d -v $(pwd):/c/Users/dev2/Projets\ web default
and then 
docker exec -it default /bin/bash

if web is your container name yes :)

try it and let us know how it goes

**Comment by tiangolo** [comment]
Hi @mysth, I think there are a couple specific issues happening to you at the same time:

---

## 1:

By the 

```
docker ps
```

resulting in

```
Get http://127.0.0.1:2375/v1.20/containers/json: dial tcp 127.0.0.1:2375: ConnectEx tcp: Aucune...
```

I think your terminal is not being able to "talk" to Docker and probably if you try to run it from the official Quickstart Terminal it won't work either. You probably ran into that issue after a sleep / wake Windows cycle, Windows detected an IP conflict with your VirtualBox Host-Only adapter and internally changed its IP, and then Docker wasn't able to connect and "talk" with the VM because there wasn't a route to get to the VM IP. More details about that and the fix are in this comment: https://github.com/docker/toolbox/issues/66#issuecomment-138755855

After you can successfully do a `docker ps` (from the official QuickStart Terminal or from Babun) we can continue with the next point.

---

## 2:

 You are running an interactive TTY command with slash symbols:

```
docker run-it -v $(pwd):/c/Users/dev2/Projets\ web default bash
```

It is interactive because you set the flags `-it`, so it will end up trying to run with winpty or it wouldn't work either way (and it wouldn't work in the official Quickstart Terminal either). If I'm not wrong, winpty does a weird thing and converts slashes to back-slashes, so your command will be interpreted as something like:

```
docker run-it -v $(pwd):\c\Users\dev2\Projets\ web default bash
```

which just won't work.

To avoid winpty converting those slashes you can try running it in daemon mode, with the `-d` flag, as @xinity suggested, the only caveat is that your container will run only while there's a process running, and the "bash" process will exit right away as it doesn't have anything to wait for. So, you have to run a process that runs "forever" to keep your container alive before doing a `docker exec...`.

So, for example, you can run a bash "script" that will just wait there sleeping until you manually kill it:

```
docker run -d -v $(pwd):/c/Users/dev2/Projets\ web --name my_container default bash -c "while true ; do sleep 1 ; done"
```

It will run a docker container in daemon mode (`-d`), it will mount a volume (`-v`) mapping the directory that you are actually in into a directory in the route `/c/Users/dev2/Projets web` inside your container (that back-slash will "escape" the space between "Projets" and "web"), it will name this container "my_container" so that you can use it later and it will run the container based on an image named "default" and it will run a "bash" script (the script goes after the `-c` flag) that will wait infinitely in a loop, sleeping for 1 second.

after that, you can do:

`docker exec -it my_container bash`

in that last command there are no slashes, so you can run it interactively (`-it`) because winpty will not break the slashes, and as the container will be already running, it will be able to "exec" another program in that container, calling it by its name, as you named it before.

---

## 3:

The last problem, you are using: `$(pwd)`, in the Quickstart Terminal it would output something like `/c/Users/mysth/code`, but in Cygwin consoles (like Babun) it will use a different path, it will output something like: `/cygdrive/c/Users/mysth/code`. To fix that last part about the volumes in Babun, follow this Wiki entry: https://github.com/tiangolo/babun-docker/wiki/Docker-Volumes-with-Babun

**Comment by tiangolo** [comment]
@mysth: I just automatized all the mounting of volumes. 

You can update **docker-babun** with `docker-babun-update`, then turn off your VirtualBox machine (if it is turned on) and use the commands as you would expect. They should "just work". At least the part about mounting volumes.

With this update all the drives in your Windows (and sub directories) are available to be used as volumes, not just those under `C:\Users` as with Docker Toolbox.

I would appreciate if you can test it and reply back with your results before closing this issue.

**Comment by lperamo** [comment]
To be honest, i'm very new to Docker and i'm a bit lost.
The post i had created was for my job but this week-end, it's for my personal use and for the context, i try to use this image zend/php:7.0. I have built this image into 

> /cygdrive/c/Users/lionel/Documents/Docker Images/webMachine/PHP7

I don't really understand what is each parameter after `-v`. What part is on my local drive, what part is on the VM ...

For your information, the first step you describe for ip addresses works as expected :)
And your babun-docker-update changes are working.

**Comment by tiangolo** [comment]
OK, @mysth, simple guide:

After the `-v` flag, the first part before the colon is the host directory, now you can just assume that it's your local / host machine directory, it will just work.

**Technical note**: (you don't need to understand this) babun-docker is mapping each directory in your local machine (your Windows) in the babun directories (`/cygdrive/c/Users/etc...`) to the directories inside your VM in the same route (`/cygdrive/c/Users/etc...`). So it will just "magically" work.

 So, that's the first part, your local directory: after the `-v` and before the colon ("`:`"). It could be: 

```
/cygdrive/c/Users/lionel/Documents/Docker Images/webMachine/PHP7
```

The second part after the colon is the directory inside your container that will have the files that are in your local directory. That could be, for example:

```
/var/www
```

So, with that, you would have a full command that looks something like:

```
docker run -d -v /cygdrive/c/Users/lionel/Documents/Docker Images/webMachine/PHP7:/var/www zend/php:7.0
```

(or something similar).

With a command like that, you could add, change or remove files inside your local directory in `/cygdrive/c/Users/lionel/Documents/Docker Images/webMachine/PHP7` and those changes would be reflected inside your container in `/var/www`, you could update a file `index.php` and the file would be updated and you could see the changes right away without needing to stop, remove and start again your container.

**Note**: Have in mind that, once you finish your development phase and you want to send to production you should not be using volumes (that thing with the `-v` flag), but rather [a Docker file with a `COPY` instruction](http://docs.docker.com/engine/reference/builder/#copy).

---

Now, the second part, the `$(pwd)` thing.

The command `pwd` **P**rints your current **W**orking **D**irectory, so, if you entered that directory described above with:

```
cd /cygdrive/c/Users/lionel/Documents/Docker Images/webMachine/PHP7
```

and then you typed:

```
pwd
```

it would print out:

```
/cygdrive/c/Users/lionel/Documents/Docker Images/webMachine/PHP7
```

The syntax `$(my_super_command)` executes a command (in this example named `my_super_command`) and returns it's results. So, it's a way to embed a command inside another.

With that, you could use a shortcut, that is used extensively in the examples of Docker, so you wouldn't have to type all this:

```
docker run -d -v /cygdrive/c/Users/lionel/Documents/Docker Images/webMachine/PHP7:/var/www zend/php:7.0
```

but rather (assuming you entered that directory already with `cd`) you could just type:

```
docker run -d -v $(pwd):/var/www zend/php:7.0
```

---

Enough explanation, here's the cheatsheet that you need.

There's no `zend/php` image, there's a [`zend/php-zendserver`](https://hub.docker.com/r/zend/php-zendserver/), but actually, there's an [official image for `php-zendserver`](https://hub.docker.com/_/php-zendserver/), so you should use it.

Go to your application directory:

```
cd /cygdrive/c/Users/lionel/Documents/Docker Images/webMachine/PHP7
```

Run your server container, with a name:

```
docker run -d --name webmachine -v $(pwd):/var/www php-zendserver
```

Enter it with `docker exec` after it is running, accessing it by its name:

```
docker exec -it webmachine bash
```

And there you have your PHP Zend server running AND a bash interactive session with all your application files dynamically updating.

**Comment by lperamo** [comment]
Ok i will testing that. The image that i have actually is :
https://github.com/janatzend/docker-php7-nightly-build

The official image is more stable I guess (i use the nightlybuild in order to have more functionalities) ?

**Comment by lperamo** [comment]
I have created the www directory in the VM. When i try to  run the container, it creates it but it can't start it.

> Error response from daemon: Cannot start container webmachine: [8] System error: no such file or directory
> Error: failed to start containers: [webmachine]

An idea ? Thank you very much for your explanations and your time so far :)

**Comment by tiangolo** [comment]
@mysth: You don't have to create a www directory in the VirtualBox VM, unless you are referring to the Docker container.

I would suggest you to run the official image unless you really need that other Github repo, it would be quite a lot easier to debug for you.

With that repo you would need to clone and build the repo, etc, something like:

```
cd /cygdrive/c/Users/lionel/Documents/
mkdir code
cd code
git clone https://github.com/janatzend/docker-php7-nightly-build.git
cd docker-php7-nightly-build
docker build -t zend/php:7.0 .
cd /cygdrive/c/Users/lionel/Documents/Docker Images/webMachine/PHP7
docker run -d --name webmachine -p 80:80 -v $(pwd):/www/public zend/php:7.0
docker exec -it webmachine bash
```

Apart from that I think I can't help much there, it seems you have errors in your container creation somewhere, by the `System error: no such file or directory`, it's possible that the "no such file or directory" is referring to a directory in your Windows or a directory inside the container, but I have no way to guess it.

If you need more help you should paste here the exact commands you are typing and the exact responses you are receiving in order. That way I could help you more easily.

**Comment by lperamo** [comment]
Yeah normally the dockerfile said that it will create this directory but when i added that directory in the VM, i believed that it has unblocked something.

Ok then I will test your suggestions tomorrow.

**Comment by lperamo** [comment]
I have executed the lines you gave to me and it works :)
I just had problems with the space in `Docker Images` but i have replaced that by the full path...waiting for better idea. I tried to use `alias pwd2="pwd | sed 's/ /\\ /g'"` and use it for the command with no success.

I don't see /var/www in the VM. Do i have to do `docker-compose` or something ?
`php -v` do nothing as well.

**Comment by tiangolo** [comment]
I guess you are having problems with the spaces in the directory in:

```
-v $(pwd):/www/public
```

in this command:

```
docker run -d --name webmachine -p 80:80 -v $(pwd):/www/public zend/php:7.0
```

You can put the whole volume part inside quotes and it will work:

```
docker run -d --name webmachine -p 80:80 -v "$(pwd):/www/public" zend/php:7.0
```

I think the image you are using doesn't use `/var/www` but `/www/public`, that's why I used that in the volume part of the `docker run` command.

About the `php -v` doing nothing, I can't help you there, that's a problem related to the very custom image you are using. Again, I would suggest you use the official image.

**Comment by lperamo** [comment]
I use this time the official image and on the VM there is no `/www/public` ... and ...same for php -v.
Am i missing something ? I see the container with `docker ps -a` so it is running but ...

**Comment by lperamo** [comment]
I have tested also

> docker run -p 88:80 -p 10088:10081 php-zendserver

And it shows ...

> Starting Zend Server 7.0.0 ..
> 
> Starting Zend Server Monitor node [OK]
> [15.11.2015 20:08:37 SYSTEM] watchdog for monitor is running.
> [15.11.2015 20:08:37 SYSTEM] monitor is running.
> - Starting web server apache2
>   AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 172.17.0.4. Set the 'ServerName' directive globally to suppress this message
>   *
>   spawn-fcgi: child spawned successfully: PID: 154
>   Starting Zend Server GUI [Lighttpd] [OK]
>   [15.11.2015 20:08:40 SYSTEM] watchdog for lighttpd is running.
>   [15.11.2015 20:08:40 SYSTEM] lighttpd is running.
>   Starting Session Clustering daemon [OK]
>   [15.11.2015 20:08:41 SYSTEM] watchdog for scd is running.
>   [15.11.2015 20:08:41 SYSTEM] scd is running.
>   Starting Deployment [OK]
>   [15.11.2015 20:08:42 SYSTEM] watchdog for zdd is running.
>   [15.11.2015 20:08:42 SYSTEM] zdd is running.
>   Starting JobQueue [OK]
>   [15.11.2015 20:08:43 SYSTEM] watchdog for jqd is running.
>   [15.11.2015 20:08:43 SYSTEM] jqd is running.
>   Starting ZendServerDaemon [OK]
>   [15.11.2015 20:08:44 SYSTEM] watchdog for zsd is running.
>   [15.11.2015 20:08:44 SYSTEM] zsd is running.
> 
> Zend Server started...
> Bootstrapping single server
> Error: failed accessing Zend Server through the Web API
> WebAPI error message : License key is invalid
> WebAPI error code    : internalServerError
> restarting ZS
> Error: failed accessing Zend Server through the Web API
> WebAPI error message : Bootstrap is needed in restartPhp
> WebAPI error code    : serverNotReady
> Zend Server is ready for use
> Your application is available at http://172.17.0.4
> To access Zend Server, navigate to http://172.17.0.4:10081
> Your admin password is MmZkNDlh
> WSARecv tcp 192.168.99.1:20898: Une tentative de connexion a échoué car le parti connecté n’a pas répondu convenablement au-delà d’une certaine durée ou une connexion établie a échoué car l’hôte de connexion n’a pas répondu.

I don't know what do exactly Zend Server but all i want to have is just an environment with PHP7, MySql, Apache... . Moreover, ZendServer is not free, isn't it ? And apparently, in addition to that, i don't succeed to make it work :/

**Comment by tiangolo** [comment]
Hehehe, all the problem all this time and you don't really need Zend Server...

Look, use the official version of PHP with Apache, read the docs:

https://hub.docker.com/_/php/

There's an example at the end to use it without Dockerfile if you want to test:

```
docker run -it --rm --name my-apache-php-app -v "$PWD":/var/www/html php:5.6-apache
```

Use an external container for your MySQL server, use the official version, read the docs:

https://hub.docker.com/_/mysql/

There are examples on how to link both containers, with something like:

```
docker run --name some-mysql -e MYSQL_ROOT_PASSWORD=my-secret-pw -d mysql:tag
```

and 

```
docker run --name some-app --link some-mysql:mysql -d application-that-uses-mysql
```

I will close this issue now as all these later problems are related to your containers, code and environment and not really to **babun-docker**.
